lib/archive/diff.go

```diff
@@ -0,0 +1,218 @@
+	ChangeTypeKey = "change_type"
+)
+
+// for sort.Sort
+type changesByPath []docker.Change
+
+func (c changesByPath) Less(i, j int) bool { return c[i].Path < c[j].Path }
+func (c changesByPath) Len() int           { return len(c) }
+func (c changesByPath) Swap(i, j int)      { c[j], c[i] = c[i], c[j] }
+
+// Diff produces a tar archive containing the differences between two filesystems
+func Diff(op trace.Operation, newDir, oldDir string, pathspec map[string]string, data bool) (io.ReadCloser, error) {
+
+	changes, err := docker.ChangesDirs(newDir, oldDir)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Sort(changesByPath(changes))
+
+	spec, err := processPathSpec(op, pathspec)
+	if err != nil {
+		return nil, err
+	}
+
+	// prune excluded paths from change set
+	for i := 0; i < len(changes); {
+		if excluded(op, changes[i].Path, spec) {
+			changes = append(changes[:i], changes[i+1:]...)
+			continue
+		}
+		i++
+	}
+
+	return Tar(op, newDir, changes, spec, data)
+}
+
+func Tar(op trace.Operation, dir string, changes []docker.Change, spec *PathSpec, data bool) (io.ReadCloser, error) {
+	var (
+		err error
+		hdr *tar.Header
+	)
+
+	// Note: it is up to the caller to handle errors and the closing of the read side of the pipe
+	r, w := io.Pipe()
+	go func() {
+		tw := tar.NewWriter(w)
+		defer func() {
+			var cerr error
+			if cerr = tw.Close(); cerr != nil {
+				op.Errorf("Error closing tar writer: %s", cerr.Error())
+			}
+			if err == nil {
+				_ = w.CloseWithError(cerr)
+			} else {
+				_ = w.CloseWithError(err)
+			}
+			return
+		}()
+
+		for _, change := range changes {
+			hdr, err = createHeader(op, dir, change, spec)
+			if err != nil {
+				op.Errorf("Error creating header from change: %s", err.Error())
+				return
+			}
+
+			var f *os.File
+
+			if !data {
+				// reset the size to 0 and include the change type
+				hdr.Size = 0
+				hdr.Xattrs = make(map[string]string)
+				hdr.Xattrs[ChangeTypeKey] = change.Kind.String()
+			}
+
+			_ = tw.WriteHeader(hdr)
+
+			p := filepath.Join(dir, change.Path)
+			if hdr.Typeflag == tar.TypeReg && hdr.Size != 0 {
+				f, err = os.Open(p)
+				if err != nil {
+					if os.IsPermission(err) {
+						err = nil
+					}
+					return
+				}
+				if f != nil {
+					_, err = io.Copy(tw, f)
+					if err != nil {
+						op.Errorf("Error writing archive data: %s", err.Error())
+					}
+					_ = f.Close()
+				}
+			}
+		}
+	}()
+	return r, err
+}
+
+// handle exclusion/inclusion of paths
+func excluded(op trace.Operation, filePath string, spec *PathSpec) bool {
+	p := filePath
+	exclude := false
+	for p != string(filepath.Separator) {
+		if _, ok := spec.Inclusions[p]; ok {
+			return false
+		}
+		if _, ok := spec.Exclusions[p]; ok {
+			exclude = true
+			break
+		}
+		p = filepath.Dir(p)
+	}
+
+	return exclude
+}
+
+func createHeader(op trace.Operation, dir string, change docker.Change, spec *PathSpec) (*tar.Header, error) {
+	var hdr *tar.Header
+	timestamp := time.Now()
+
+	switch change.Kind {
```

Comment:
FYI, this looks like it will cause a problem with checksum verification. Currently, we hash the entire tarstream (not just the actual data). This appears to be how docker does it in `applyLayer` (which we use in `WriteImage`). Having the change attrs in the stream (among other differences in the headers) results in a different hash.